### PR TITLE
Idiomatic errors

### LINKED ISSUE #823: Idiomatic errors
### Problem to Solve

We use static strings as errors in `core` crates which is not idiomatic and doesn't allow us to forward through syscalls errors using codes

### Possible Solution

Introduce `core-errors` crate which contains only errors to be thin and might be used in `core-processor`, `gcore` crates

### Notes

_No response_

### PR DESCRIPTION
**Pull request by ark0f**
Resolves #823.

Also get rid of `*_TRAP_STR` constants

### CODE REVIEW

File: core-backend/common/src/funcs.rs [hunk 1]
```diff
@@ -16,44 +16,28 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::{GAS_ALLOWANCE_STR, LEAVE_TRAP_STR, WAIT_TRAP_STR};
 use alloc::{vec, vec::Vec};
 use gear_core::memory::Memory;
+use gear_core_errors::MemoryError;
```
Comment (SabaunT) [inline core-backend/common/src/funcs.rs:21]:
```suggestion
use gear_core::memory::{Memory, Error as MemoryError};
```

**Review by SabaunT — CHANGES_REQUESTED** [review]
Interim review

File: common/src/lazy_pages.rs [hunk 2]
```diff
@@ -20,20 +20,52 @@
...
+    fn from(err: MprotectError) -> Self {
+        Self::Mprotect(err)
+    }
+}
+
+impl From<GetReleasedPageError> for Error {
+    fn from(err: GetReleasedPageError) -> Self {
+        Self::GetReleasedPage(err)
+    }
+}
```
Comment (SabaunT) [inline common/src/lazy_pages.rs:58]: what about using `thiserror` crate. It really helps with removing this.

File: runtime-interface/src/lib.rs [hunk 3]
```diff
@@ -21,22 +21,38 @@
...
 
-#[derive(Debug, Encode, Decode)]
+impl fmt::Display for MprotectError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MprotectError::PageError => write!(f, "Page error"),
+            MprotectError::OsError => write!(f, "OS error"),
+        }
+    }
+}
```
Comment (SabaunT) [inline runtime-interface/src/lib.rs:45]:
```
#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, derive_more::Display)]
pub enum MprotectError {
    #[display(fmt = "Page error")]
    PageError,
    #[display(fmt = "OS error")]
    OsError,
}
```

File: runtime-interface/src/lib.rs [hunk 4]
```diff
@@ -21,22 +21,38 @@
...
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
 pub struct GetReleasedPageError;
 
+impl fmt::Display for GetReleasedPageError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Failed to get released page")
+    }
+}
```
Comment (SabaunT) [inline runtime-interface/src/lib.rs:54]: same as with the previous error

**Review by SabaunT — CHANGES_REQUESTED** [review]

File: core-backend/common/src/funcs.rs (hunk 1 above)
Comment (ark0f) [reply core-backend/common/src/funcs.rs:21]: Is this so important?

File: common/src/lazy_pages.rs (hunk 2 above)
Comment (ark0f) [reply common/src/lazy_pages.rs:58]: `common` crate uses `no_std` when `thiserror` relies on `std::error::Error`

Comment (SabaunT) [reply common/src/lazy_pages.rs:58]: we can still remove that `Display` implementation with `#[derive(fmt = "")]`, isn't it?

File: core-backend/common/src/funcs.rs (hunk 1 above)
Comment (SabaunT) [reply core-backend/common/src/funcs.rs:21]: Of course, in comparison to significant changes that you did, this is minor stuff. During review we see minor/major things to be changed. This is minor stuff and you can leave it how it is. Or you can emphasize using re-export. :)

File: core/src/env.rs [hunk 5]
```diff
@@ -173,19 +181,17 @@ impl<E: Ext> ExtCarrier<E> {
     }
 
     /// Calls infallible fn with inner ext.
-    pub fn with<R>(&self, f: impl FnOnce(&mut E) -> R) -> Result<R, &'static str> {
+    pub fn with<T>(&self, f: impl FnOnce(&mut E) -> T) -> Result<T, LaterExtWithError> {
```
[the remark this answers is not in the document]
Comment (ark0f) [reply core/src/env.rs]:
```
error[E0283]: type annotations needed
   --> core-backend\wasmtime\src\funcs.rs:627:17
    |
627 |             ext.with_fallible(|ext: &mut E| Ok(ext.msg().len() as _))
    |             ----^^^^^^^^^^^^^----------------------------------------
    |             |   |
    |             |   cannot infer type for type parameter `U` declared on the associated function `with_fallible`
    |             this method call resolves to `core::result::Result<T, U>`
    |
    = note: cannot satisfy `_: core::convert::From<gear_core::env::ExtCarrierWithError>`
```
So I have to write `ext.with_fallible::<_, ExtCarrierWithError>(...)` which is quite ugly

File: core-backend/wasmtime/src/env.rs [hunk 6]
```diff
@@ -299,22 +332,16 @@ impl<E: Ext + IntoExtInfo> Environment<E> for WasmtimeEnvironment<E> {
 
         let res = entry_func.call(&mut self.store, &[], &mut []);
 
+        let termination_reason = self.store.data().termination_reason.clone();
```
Comment (ark0f) [inline core-backend/wasmtime/src/env.rs:335]: Found bug - `TerminationReason::GasAllowanceExceeded` from `Ext::charge_gas*` is not considered

File: core/src/env.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (SabaunT) [reply core/src/env.rs]: ok!

**Review by SabaunT — APPROVED** [review]

File: core-backend/common/src/funcs.rs (hunk 1 above)
Comment (ark0f) [reply core-backend/common/src/funcs.rs:21]: > During review I see minor/major things to be improved

Don't be afraid to say about them. I just wondering sometimes

File: core-backend/wasmtime/src/env.rs (hunk 6 above)
Comment (ark0f) [reply core-backend/wasmtime/src/env.rs:335]: Considered actually

File: core/src/memory.rs [hunk 7]
```diff
@@ -196,7 +165,7 @@ impl core::ops::Sub for WasmPageNumber {
 /// ```
 pub fn pages_to_wasm_pages_set<'a>(
     pages_iter: impl Iterator<Item = &'a PageNumber>,
-) -> Result<BTreeSet<WasmPageNumber>, &'static str> {
+) -> Result<BTreeSet<WasmPageNumber>, MemoryError> {
```
Comment (gshep) [inline core/src/memory.rs:168]: why not use alias `Error` and remove the second use?

File: core/src/memory.rs [hunk 8]
```diff
@@ -205,7 +174,7 @@ pub fn pages_to_wasm_pages_set<'a>(
                 wasm_pages.insert(WasmPageNumber(gp.0 / PageNumber::num_in_one_wasm_page()));
                 Ok(())
             } else {
-                Err("There is wasm page, which has not all gear pages in the begin")
+                Err(MemoryError::NotAllPagesInEnd)
```
Comment (gshep) [inline core/src/memory.rs:177]: Maybe something like `NotAllPagesInBegin` should be used? The error message states `... in the begin`

**Review by gshep — APPROVED** [review]
Good job! :hammer_and_wrench:

File: common/src/lazy_pages.rs [hunk 9]
```diff
@@ -120,7 +141,9 @@ pub fn protect_lazy_pages_and_update_wasm_mem_addr(
         );
         gear_ri::set_wasm_mem_begin_addr(new_mem_addr);
     }
-    mprotect_lazy_pages(new_mem_addr, true)
+    mprotect_lazy_pages(new_mem_addr, true)?;
```
[the remark this answers is not in the document]
Comment (ark0f) [reply common/src/lazy_pages.rs:144]: Why?

[the remark this answers is not in the document]
Comment (gshep) [reply common/src/lazy_pages.rs:144]: Return type is the same

**Review by shamilsan — APPROVED** [review]
👍

**Comment by ark0f** [comment]
@shamilsan, why no `C2-refactoring` label?

**Comment by shamilsan** [comment]
> @shamilsan, why no `C2-refactoring` label?

We use these labels in issues only, not in PRs.

**Pull request merged**